src/source/from-shift.js

```diff
@@ -0,0 +1,204 @@
+
+	let contents;
+
+	switch (kind) {
+		case 'FunctionDeclaration':
+		case 'Method':
+			contents = {
+				type: 'FunctionOrMethodContents',
+				isThisCaptured: false
+			};
+			break;
+		case 'FunctionExpression':
+			contents = {
+				type: 'FunctionExpressionContents',
+				isFunctionNameCaptured: false,
+				isThisCaptured: false
+			};
+			break;
+		case 'ArrowExpression':
+			if (isExpressionBody) {
+				contents = { type: 'ArrowExpressionContentsWithExpression' };
+				obj.type = 'ArrowExpressionWithExpression';
+			} else {
+				contents = { type: 'ArrowExpressionContentsWithFunctionBody' };
+				obj.type = 'ArrowExpressionWithFunctionBody';
+			}
+			break;
+		case 'Getter':
+			contents = {
+				type: 'GetterContents',
+				isThisCaptured: false
+			};
+			break;
+		case 'Setter': {
+			let { param } = obj;
+			delete obj.param;
+			let { parameterScope, length } = parameterScopeAndLength([param]);
+			contents = {
+				type: 'SetterContents',
+				isThisCaptured: false,
+				param,
+				parameterScope
+			};
+			obj.length = length;
+			break;
+		}
+		default:
+			throw new TypeError(`Unexpected function kind ${kind}`);
+	}
+
+	if (kind !== 'Getter' && kind !== 'Setter') {
+		let { params } = obj;
+		delete obj.params;
+
+		contents.params = params;
+
+		let { parameterScope, length } = parameterScopeAndLength(params.items);
+
+		contents.parameterScope = parameterScope;
+		obj.length = length;
+		obj.isAsync = !!obj.isAsync;
+	}
+
+	contents.bodyScope = dummyDeclaredScope('AssertedVarScope');
+	contents.body = body;
+
+	obj.contents = contents;
+	obj.directives = directives;
+
+	makeEager(obj);
+}
+
+function convertObject(obj) {
+	switch (obj.type) {
+		case 'BlockStatement':
+			obj = obj.block;
+			// fallthrough to handle the unwrapped `Block` node
+		case 'Block':
+			obj.scope = dummyDeclaredScope('AssertedBlockScope');
+			break;
+		case 'ForInStatement':
+		case 'ForOfStatement': {
+			let { left } = obj;
+			if (left.type === 'VariableDeclaration') {
+				left.type = 'ForInOfBinding';
+				left.binding = left.declarators[0].binding;
+				delete left.declarators;
+			}
+			break;
+		}
+		case 'FunctionDeclaration':
+		case 'Method':
+		case 'FunctionExpression':
+		case 'ArrowExpression':
+		case 'Getter':
+		case 'Setter':
+			createFunctionContents(obj);
+			break;
+		case 'LabeledStatement':
+			obj.type = 'LabelledStatement';
+			break;
+		case 'LiteralRegExpExpression': {
+			let flags = '';
+			if (obj.global) flags += 'g';
+			if (obj.ignoreCase) flags += 'i';
+			if (obj.multiLine) flags += 'm';
+			if (obj.sticky) flags += 'y';
+			if (obj.unicode) flags += 'u';
+			obj.flags = flags;
+			break;
+		}
+		case 'Script':
+			obj.scope = dummyDeclaredScope('AssertedScriptGlobalScope');
+			break;
+		case 'CatchClause':
+			obj.bindingScope = dummyBoundNamesScope();
+			break;
+		case 'StaticPropertyName':
+			obj.type = 'LiteralPropertyName';
+			break;
+		case 'VariableDeclarationStatement':
+			obj = obj.declaration;
```

Comment:
this is doing the same thing as the BlockStatement/Block.
it would be better leaving some comment that `VariableDeclaration` here won't be converted.